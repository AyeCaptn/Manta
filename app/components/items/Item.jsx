// Libs
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

// Custom Components
import { TR, TD } from '../shared/Table';
import Button from '../shared/Button';

// Component
class Item extends PureComponent {
  constructor(props) {
    super(props);
    this.deleteItem = this.deleteItem.bind(this);
    this.editItem = this.editItem.bind(this);
  }

  deleteItem() {
    const { deleteItem, item } = this.props;
    deleteItem(item._id);
  }

  editItem() {
    const { editItem, item } = this.props;
    editItem(item);
  }

  render() {
    const { item } = this.props;
    return (
      <TR>
        <TD bold>{item.description}</TD>
        <TD>{item.price}</TD>
        <TD actions>
          <Button link primary onClick={this.editItem}>
            <i className="ion-edit" />
          </Button>
          <Button link danger onClick={this.deleteItem}>
            <i className="ion-close-circled" />
          </Button>
        </TD>
      </TR>
    );
  }
}

Item.propTypes = {
  item: PropTypes.object.isRequired,
  deleteItem: PropTypes.func.isRequired,
  editItem: PropTypes.func.isRequired,
};

export default Item;
