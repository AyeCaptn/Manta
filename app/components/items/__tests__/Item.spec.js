// Libs
import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

// Component
import Item from '../Item';
import { TR, TD } from '../../shared/Table';
import Button from '../../shared/Button';

// Mocks
const item = {
  _id: 'id-string',
  name: 'Product/Service A',
  description: 'Description of the product/service',
  price: 12.3,
};
const editItem = jest.fn();
const deleteItem = jest.fn();

describe('Renders correctly to the DOM', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Item
        item={item}
        editItem={editItem}
        deleteItem={deleteItem}
      />
    );
  });

  it('renders 1 TR row', () => {
    expect(wrapper.find(TR)).toHaveLength(1);
    expect(wrapper.find(TR)).not.toHaveLength(2);
  });

  it('renders 5 TD columns', () => {
    expect(wrapper.find(TD)).toHaveLength(5);
    expect(wrapper.find(TD)).not.toHaveLength(6);
  });

  it('receive correct props', () => {
    const mountWrapper = mount(
      <Item
        item={item}
        editItem={editItem}
        deleteItem={deleteItem}
      />
    );
    expect(mountWrapper.prop('item')).toEqual(item);
    expect(mountWrapper.prop('editItem')).toEqual(editItem);
    expect(mountWrapper.prop('deleteItem')).toEqual(deleteItem);
  });

  it('renders 2 buttons in the last TD', () => {
    expect(
      wrapper
        .find(TD)
        .last()
        .find(Button)
    ).toHaveLength(2);
    expect(
      wrapper
        .find(TD)
        .last()
        .find(Button)
    ).not.toHaveLength(3);
  });

  describe('renders buttons correctly', () => {
    let firstBtn, lastBtn;
    beforeEach(() => {
      firstBtn = wrapper.find(Button).first();
      lastBtn = wrapper.find(Button).last();
    });

    it('renders buttons with correct props', () => {
      expect(firstBtn.prop('link')).toBe(true);
      expect(firstBtn.prop('primary')).toBe(true);
      expect(firstBtn.prop('success')).toBe(false);
      expect(lastBtn.prop('link')).toBe(true);
      expect(lastBtn.prop('danger')).toBe(true);
      expect(lastBtn.prop('success')).toBe(false);
    });

    it('handle click correctly', () => {
      firstBtn.simulate('click');
      expect(editItem).toHaveBeenCalled();
      expect(editItem).toHaveBeenCalledWith(item);

      lastBtn.simulate('click');
      expect(deleteItem).toHaveBeenCalled();
      expect(deleteItem).toHaveBeenCalledWith('id-string');
    });
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Item
          item={item}
          editItem={editItem}
          deleteItem={deleteItem}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
