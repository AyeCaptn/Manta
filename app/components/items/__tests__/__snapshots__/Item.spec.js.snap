// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly to the DOM matches snapshot 1`] = `
<tr
  className="Table__TRStyle-gHmqLx BngTM"
>
  <td
    className="Table__TDStyle-cAmfHP hlnQVD"
  >
    Description of the product/service
  </td>
  <td
    className="Table__TDStyle-cAmfHP jrRCak"
  >
    12.3
  </td>
  <td
    className="Table__TDStyle-cAmfHP evw"
  >
    <button
      className="Button__ButtonLinkStyle-caVwCA lhMtRH"
      onClick={[Function]}
    >
      <i
        className="ion-edit"
      />
    </button>
    <button
      className="Button__ButtonLinkStyle-caVwCA gvxXrI"
      onClick={[Function]}
    >
      <i
        className="ion-close-circled"
      />
    </button>
  </td>
</tr>
`;
